File: client/src/voice/protocol.ts
// Wire protocol for the WebRTC DataChannel between the phone and
// daemon. Mirror of `daemon/src/protocol.ts`; the protocol test pins
// the two copies to the same serialized shape.
//
// Routing (sessionId, delivery channel/target) is bound once at
// rendezvous when the per-session voice room is created. `stt.start`
// no longer carries routing per turn. Voice settings (legacy voice id plus
// canonical TTS provider/model/voice selection) flow over the voice room:
// an initial value is included in
// `rendezvous.join` so the first reply uses it, and `settings.update`
// applies subsequent changes without reconnecting. The phone can request
// the daemon's current TTS catalog over the same channel.

export interface DeliveryTarget {
  channel: string;
  target?: string;
}

export interface RendezvousJoinInput {
  sessionId: string;
  delivery?: DeliveryTarget;
}

export interface TtsSelection {
  providerId?: string;
  model?: string;
  voice?: string;
}

export interface SttSelection {
  providerId?: string;
  model?: string;
}

export interface VoiceSettings {
  voice?: string;
  tts?: TtsSelection;
  stt?: SttSelection;
}

export interface TtsCatalogVoice {
  id: string;
  name: string;
}

export interface TtsCatalogProvider {
  id: string;
  name: string;
  configured: boolean;
  selected: boolean;
  available: boolean;
  models: string[];
  voices: TtsCatalogVoice[];
}

export interface TtsCatalog {
  activeProvider?: string;
  generatedAt: string;
  providers: TtsCatalogProvider[];
}

export interface SttCatalogProvider {
  id: string;
  name: string;
  configured: boolean;
  selected: boolean;
  available: boolean;
  models: string[];
}

export interface SttCatalog {
  activeProvider?: string;
  generatedAt: string;
  providers: SttCatalogProvider[];
}

export type PhoneToDaemon =
  | {
      t: 'rendezvous.join';
      sessionId: string;
      delivery?: DeliveryTarget;
      settings?: VoiceSettings;
    }
  | { t: 'settings.update'; settings: VoiceSettings }
  | { t: 'tts.catalog.request' }
  | { t: 'stt.catalog.request' }
  | { t: 'stt.start' }
  | { t: 'stt.audio.done' }
  | { t: 'stt.cancel' }
  | { t: 'reply.cancel' };

export type DaemonToPhone =
  | { t: 'rendezvous.accept'; roomId: string }
  | { t: 'rendezvous.error'; message: string }
  | { t: 'session.replaced'; reason: string }
  | { t: 'stt.ready' }
  | { t: 'stt.partial'; text: string; is_final: boolean }
  | { t: 'stt.done'; text: string }
  | { t: 'stt.error'; message: string }
  | { t: 'stt.closed' }
  | { t: 'reply.start'; text: string }
  | { t: 'reply.done'; text: string }
  | { t: 'reply.error'; message: string }
  | { t: 'tts.start'; sample_rate: number }
  | { t: 'tts.catalog'; catalog: TtsCatalog }
  | { t: 'stt.catalog'; catalog: SttCatalog }
  | { t: 'tts.done' }
  | { t: 'tts.error'; message: string };

export const phoneToDaemon = {
  rendezvousJoin: (input: RendezvousJoinInput & { settings?: VoiceSettings }): PhoneToDaemon => ({
    t: 'rendezvous.join',
    sessionId: input.sessionId,
    ...(input.delivery ? { delivery: input.delivery } : {}),
    ...(input.settings ? { settings: input.settings } : {}),
  }),
  settingsUpdate: (settings: VoiceSettings): PhoneToDaemon => ({
    t: 'settings.update',
    settings,
  }),
  ttsCatalogRequest: (): PhoneToDaemon => ({ t: 'tts.catalog.request' }),
  sttCatalogRequest: (): PhoneToDaemon => ({ t: 'stt.catalog.request' }),
  sttStart: (): PhoneToDaemon => ({ t: 'stt.start' }),
  sttAudioDone: (): PhoneToDaemon => ({ t: 'stt.audio.done' }),
  sttCancel: (): PhoneToDaemon => ({ t: 'stt.cancel' }),
  replyCancel: (): PhoneToDaemon => ({ t: 'reply.cancel' }),
};

export const daemonToPhone = {
  rendezvousAccept: (roomId: string): DaemonToPhone => ({ t: 'rendezvous.accept', roomId }),
  rendezvousError: (message: string): DaemonToPhone => ({ t: 'rendezvous.error', message }),
  sessionReplaced: (reason = 'newer_phone_connected'): DaemonToPhone => ({
    t: 'session.replaced',
    reason,
  }),
  sttReady: (): DaemonToPhone => ({ t: 'stt.ready' }),
  sttPartial: (text: string, isFinal: boolean): DaemonToPhone => ({
    t: 'stt.partial',
    text,
    is_final: isFinal,
  }),
  sttDone: (text: string): DaemonToPhone => ({ t: 'stt.done', text }),
  sttError: (message: string): DaemonToPhone => ({ t: 'stt.error', message }),
  sttClosed: (): DaemonToPhone => ({ t: 'stt.closed' }),
  replyStart: (text: string): DaemonToPhone => ({ t: 'reply.start', text }),
  replyDone: (text: string): DaemonToPhone => ({ t: 'reply.done', text }),
  replyError: (message: string): DaemonToPhone => ({ t: 'reply.error', message }),
  ttsStart: (sampleRate: number): DaemonToPhone => ({
    t: 'tts.start',
    sample_rate: sampleRate,
  }),
  ttsCatalog: (catalog: TtsCatalog): DaemonToPhone => ({ t: 'tts.catalog', catalog }),
  sttCatalog: (catalog: SttCatalog): DaemonToPhone => ({ t: 'stt.catalog', catalog }),
  ttsDone: (): DaemonToPhone => ({ t: 'tts.done' }),
  ttsError: (message: string): DaemonToPhone => ({ t: 'tts.error', message }),
};

File: android/app/src/main/java/app/clawkietalkie/protocol/Protocol.kt
package app.clawkietalkie.protocol

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray

// Wire protocol for the WebRTC DataChannel between the phone and daemon.
// Mirror of the web client's `src/voice/protocol.ts` (which itself mirrors
// `daemon/src/protocol.ts`). Control messages are loosely-typed JSON objects
// with a `t` discriminator; unknown fields must be tolerated so a newer
// client keeps working against an older daemon.

const val PROTOCOL_VERSION = 1

object ProtocolFeatures {
    const val TTS_CATALOG = "tts.catalog"
    const val STT_CATALOG = "stt.catalog"
    const val SESSIONS_LIST = "sessions.list"
    const val SESSIONS_CATALOG = "sessions.catalog"
}

val CLIENT_WANTED_PROTOCOL_FEATURES = listOf(
    ProtocolFeatures.TTS_CATALOG,
    ProtocolFeatures.STT_CATALOG,
    ProtocolFeatures.SESSIONS_LIST,
    ProtocolFeatures.SESSIONS_CATALOG,
)

fun isDaemonSupportedProtocol(protocol: Int?): Boolean = protocol == PROTOCOL_VERSION

val protocolJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = false
}

/**
 * A decoded control message. Equivalent to the web client's
 * `ControlMessage { t: string; [key: string]: unknown }`.
 */
class ControlMessage(val fields: JsonObject) {
    val t: String = (fields["t"] as? JsonPrimitive)?.takeIf { it.isString }?.content ?: ""

    fun string(key: String): String? =
        (fields[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

    fun number(key: String): Double? =
        (fields[key] as? JsonPrimitive)?.content?.toDoubleOrNull()?.takeIf { it.isFinite() }

    fun int(key: String): Int? = number(key)?.toInt()

    fun boolean(key: String): Boolean? =
        (fields[key] as? JsonPrimitive)?.takeIf { !it.isString }?.let {
            runCatching { it.boolean }.getOrNull()
        }

    fun obj(key: String): JsonObject? = fields[key] as? JsonObject

    fun array(key: String): JsonArray? = fields[key] as? JsonArray

    fun encode(): String = protocolJson.encodeToString(JsonObject.serializer(), fields)

    companion object {
        fun parse(text: String): ControlMessage? = runCatching {
            val element = protocolJson.parseToJsonElement(text)
            if (element is JsonObject) ControlMessage(element) else null
        }.getOrNull()

        fun of(builder: kotlinx.serialization.json.JsonObjectBuilder.() -> Unit): ControlMessage =
            ControlMessage(buildJsonObject(builder))
    }
}

// ---------------------------------------------------------------------------
// Typed payloads (decoded from control messages)
// ---------------------------------------------------------------------------

data class DeliveryTarget(
    val channel: String,
    val target: String,
    val accountId: String? = null,
)

data class RendezvousJoinInput(
    val sessionId: String,
    val sessionKey: String? = null,
    val channel: String? = null,
    val target: String? = null,
    val accountId: String? = null,
    val delivery: DeliveryTarget? = null,
)

data class TtsSelection(
    val providerId: String? = null,
    val model: String? = null,
    val voice: String? = null,
)

data class SttSelection(
    val providerId: String? = null,
    val model: String? = null,
)

data class VoiceSettings(
    val voice: String? = null,
    val tts: TtsSelection? = null,
    val stt: SttSelection? = null,
) {
    val isEmpty: Boolean
        get() = voice == null && tts == null && stt == null
}

data class TtsCatalogVoice(val id: String, val name: String)

data class TtsCatalogProvider(
    val id: String,
    val name: String,
    val configured: Boolean,
    val selected: Boolean,
    val available: Boolean,
    val models: List<String>,
    val voices: List<TtsCatalogVoice>,
)

data class TtsCatalog(
    val activeProvider: String?,
    val generatedAt: String,
    val providers: List<TtsCatalogProvider>,
)

data class SttCatalogProvider(
    val id: String,
    val name: String,
    val configured: Boolean,
    val selected: Boolean,
    val available: Boolean,
    val models: List<String>,
)

data class SttCatalog(
    val activeProvider: String?,
    val generatedAt: String,
    val providers: List<SttCatalogProvider>,
)

data class RecentSession(
    val sessionId: String,
    val sessionKey: String,
    val agent: String,
    val channel: String? = null,
    val target: String? = null,
    val accountId: String? = null,
    val lastActivity: String? = null,
    val displayLabel: String,
    val lastMessagePreview: String? = null,
    val lastMessageRole: String? = null,
    val lastAssistantPreview: String? = null,
)

data class RecentSessionsSnapshot(
    val generatedAt: String,
    val sessions: List<RecentSession>,
)

// ---------------------------------------------------------------------------
// Phone → daemon message builders (mirror of `phoneToDaemon` in protocol.ts)
// ---------------------------------------------------------------------------

object PhoneToDaemon {
    fun clientHello(wants: List<String> = CLIENT_WANTED_PROTOCOL_FEATURES): ControlMessage =
        ControlMessage.of {
            put("t", "client.hello")
            put("protocol", PROTOCOL_VERSION)
            putJsonArray("wants") { wants.forEach { add(JsonPrimitive(it)) } }
        }

    fun rendezvousJoin(input: RendezvousJoinInput, settings: VoiceSettings?): ControlMessage =
        ControlMessage.of {
            put("t", "rendezvous.join")
            put("sessionId", input.sessionId)
            input.sessionKey?.let { put("sessionKey", it) }
            input.channel?.let { put("channel", it) }
            input.target?.let { put("target", it) }
            input.accountId?.let { put("accountId", it) }
            input.delivery?.let { delivery ->
                put("delivery", buildJsonObject {
                    put("channel", delivery.channel)
                    put("target", delivery.target)
                    delivery.accountId?.let { put("accountId", it) }
                })
            }
            if (settings != null && !settings.isEmpty) put("settings", voiceSettingsJson(settings))
        }

    fun settingsUpdate(settings: VoiceSettings): ControlMessage = ControlMessage.of {
        put("t", "settings.update")
        put("settings", voiceSettingsJson(settings))
    }

    fun ttsCatalogRequest(): ControlMessage = simple("tts.catalog.request")
    fun sttCatalogRequest(): ControlMessage = simple("stt.catalog.request")
    fun sessionsListRequest(): ControlMessage = simple("sessions.list.request")
    fun sessionsCatalogRequest(): ControlMessage = simple("sessions.catalog.request")
    fun sessionsListSubscribe(): ControlMessage = simple("sessions.list.subscribe")
    fun sessionsListUnsubscribe(): ControlMessage = simple("sessions.list.unsubscribe")
    fun sttStart(): ControlMessage = simple("stt.start")
    fun sttAudioDone(): ControlMessage = simple("stt.audio.done")
    fun sttCancel(): ControlMessage = simple("stt.cancel")
    fun replyCancel(): ControlMessage = simple("reply.cancel")

    private fun simple(t: String): ControlMessage = ControlMessage.of { put("t", t) }

    fun voiceSettingsJson(settings: VoiceSettings): JsonObject = buildJsonObject {
        settings.voice?.let { put("voice", it) }
        settings.tts?.let { tts ->
            put("tts", buildJsonObject {
                tts.providerId?.let { put("providerId", it) }
                tts.model?.let { put("model", it) }
                tts.voice?.let { put("voice", it) }
            })
        }
        settings.stt?.let { stt ->
            put("stt", buildJsonObject {
                stt.providerId?.let { put("providerId", it) }
                stt.model?.let { put("model", it) }
            })
        }
    }
}

// ---------------------------------------------------------------------------
// Decoders for daemon → phone payloads
// ---------------------------------------------------------------------------

private fun JsonElement?.stringOrNull(): String? =
    (this as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun JsonElement?.booleanOrFalse(): Boolean =
    (this as? JsonPrimitive)?.let { runCatching { it.boolean }.getOrNull() } ?: false

private fun JsonElement?.stringList(): List<String> =
    (this as? JsonArray)?.mapNotNull { it.stringOrNull() } ?: emptyList()

fun decodeTtsCatalog(catalog: JsonObject): TtsCatalog {
    val providers = (catalog["providers"] as? JsonArray)?.mapNotNull { entry ->
        val source = entry as? JsonObject ?: return@mapNotNull null
        val id = source["id"].stringOrNull() ?: return@mapNotNull null
        TtsCatalogProvider(
            id = id,
            name = source["name"].stringOrNull() ?: "",
            configured = source["configured"].booleanOrFalse(),
            selected = source["selected"].booleanOrFalse(),
            available = source["available"].booleanOrFalse(),
            models = source["models"].stringList(),
            voices = (source["voices"] as? JsonArray)?.mapNotNull { voiceEntry ->
                val voice = voiceEntry as? JsonObject ?: return@mapNotNull null
                val voiceId = voice["id"].stringOrNull() ?: return@mapNotNull null
                TtsCatalogVoice(id = voiceId, name = voice["name"].stringOrNull() ?: "")
            } ?: emptyList(),
        )
    } ?: emptyList()
    return TtsCatalog(
        activeProvider = catalog["activeProvider"].stringOrNull(),
        generatedAt = catalog["generatedAt"].stringOrNull() ?: "",
        providers = providers,
    )
}

fun decodeSttCatalog(catalog: JsonObject): SttCatalog {
    val providers = (catalog["providers"] as? JsonArray)?.mapNotNull { entry ->
        val source = entry as? JsonObject ?: return@mapNotNull null
        val id = source["id"].stringOrNull() ?: return@mapNotNull null
        SttCatalogProvider(
            id = id,
            name = source["name"].stringOrNull() ?: "",
            configured = source["configured"].booleanOrFalse(),
            selected = source["selected"].booleanOrFalse(),
            available = source["available"].booleanOrFalse(),
            models = source["models"].stringList(),
        )
    } ?: emptyList()
    return SttCatalog(
        activeProvider = catalog["activeProvider"].stringOrNull(),
        generatedAt = catalog["generatedAt"].stringOrNull() ?: "",
        providers = providers,
    )
}

fun decodeRecentSession(value: JsonElement): RecentSession? {
    val source = value as? JsonObject ?: return null
    val sessionId = source["sessionId"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    val sessionKey = source["sessionKey"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    return RecentSession(
        sessionId = sessionId,
        sessionKey = sessionKey,
        agent = source["agent"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() } ?: "unknown",
        channel = source["channel"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        target = source["target"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        accountId = source["accountId"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        lastActivity = source["lastActivity"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        displayLabel = source["displayLabel"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() }
            ?: (source["agent"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() } ?: "unknown"),
        lastMessagePreview = source["lastMessagePreview"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        lastMessageRole = source["lastMessageRole"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
        lastAssistantPreview = source["lastAssistantPreview"].stringOrNull()?.trim()?.takeIf { it.isNotEmpty() },
    )
}
